driver-core/src/main/com/mongodb/internal/connection/DefaultConnectionPool.java

```diff
@@ -101,135 +119,104 @@ public InternalConnection get() {
-                        } else {
-                            PooledConnection connection = getPooledConnection(getRemainingWaitTime(), MILLISECONDS);
-                            openAsync(connection, errHandlingCallback);
-                        }
-                    } catch (MongoTimeoutException e) {
-                        Exception exception = new MongoTimeoutException(format("Timeout waiting for a pooled connection after %d %s",
-                                settings.getMaxWaitTime(MILLISECONDS), MILLISECONDS));
-                        emitCheckOutFailedEvent(exception);
-                        errHandlingCallback.onResult(null, exception);
-                    } catch (Throwable t) {
-                        emitCheckOutFailedEvent(t);
-                        errHandlingCallback.onResult(null, t);
-                    }
+            getAsyncGetter().execute(() -> {
+                if (timeout.expired()) {
+                    eventSendingCallback.onResult(null, createTimeoutException(timeout));
+                    return;
                 }
-
-                private long getRemainingWaitTime() {
-                    return startTimeMillis + settings.getMaxWaitTime(MILLISECONDS) - System.currentTimeMillis();
+                PooledConnection connection;
+                try {
+                    connection = getPooledConnection(timeout);
+                } catch (RuntimeException e) {
+                    eventSendingCallback.onResult(null, e);
+                    return;
                 }
+                openAsync(connection, timeout, SAME_THREAD_EXECUTOR, eventSendingCallback);
             });
         }
     }
 
-    private void emitCheckOutFailedEvent(final Throwable t) {
+    /**
+     * Sends {@link ConnectionCheckOutFailedEvent}
+     * and returns a {@link RuntimeException} that is reported as the cause of a checkout failure.
+     */
+    private RuntimeException checkOutFailed(final Throwable t) {
+        Throwable result = t;
         if (t instanceof MongoTimeoutException) {
             connectionPoolListener.connectionCheckOutFailed(new ConnectionCheckOutFailedEvent(serverId, Reason.TIMEOUT));
+        } else if (t instanceof MongoOpenConnectionInternalException) {
+            connectionPoolListener.connectionCheckOutFailed(new ConnectionCheckOutFailedEvent(serverId, Reason.CONNECTION_ERROR));
+            result = assertNotNull(t.getCause());
         } else if (t instanceof IllegalStateException && t.getMessage().equals("The pool is closed")) {
             connectionPoolListener.connectionCheckOutFailed(new ConnectionCheckOutFailedEvent(serverId, Reason.POOL_CLOSED));
         } else {
             connectionPoolListener.connectionCheckOutFailed(new ConnectionCheckOutFailedEvent(serverId, Reason.UNKNOWN));
         }
+        return result instanceof RuntimeException ? (RuntimeException) result : new RuntimeException(result);
```

Comment:
`SingleResultCallback.onResult(T result, Throwable t)` accepts `Throwable`, though according to the logic you expressed it should be accepting `RuntimeException`. The best answer to the question I can give is: `result` may not be an instance of `RuntimeException` in the scenarios in which not a `RuntimeException` is passed to `onResult`.

Interestingly, `CompletionHandler.failed(Throwable exc, A attachment)` used in `AsynchronousSocketChannel.read` also accepts a `Throwable` instead of accepting an `Exception`, which I fail to find a justification for. Did this fact affected the signature of our callbacks?

Your question made me realize, that `result` may generally speaking also be an `Error`, and I do not handle this. Though if this is the case, then the code that completed the callback must have violated the recommendation from the `Error` class to not catch those. I see three ways of how I can proceed:

1. [handle `Throwable`s] Also handle situations when `result` is an `Error` (we must not wrap it into an `Exception` when handling). This approach works around (not a good thing in this case) a potential problem (code catching `Error`s and completing callbacks with them) which we may not have. But if we do have it, then this approach further hides the problem.
2. [handle only `Exception`s] Leave the idea mostly as is (handles `Exception`s but not `Error`s), do the following modifications
   2.1. wrap checked exceptions in `MongoException` instead of wrapping in `RuntimeException`;
   2.2. add `assertFalse(result instanceof Error)` to make the assumption about `Error`s explicit.
3. [handle only `RuntimeException`s] Add `assertTrue(result instanceof RuntimeException)` and then treat `result` as a `RuntimeException`, wrapping it in `MongoException` if needed.

I reject the option 1. I think that the option 2 is the least incorrect one, but I will not argue against the option 3. @jyemin Do you prefer 2 or 3, or maybe you see an even better approach that I missed?

Reply:
`SingleResultCallback.onResult(T result, Throwable t)` accepts `Throwable`, though according to the logic you expressed it should be accepting `RuntimeException`. The best answer to the question I can give is: `result` may not be an instance of `RuntimeException` in the scenarios in which not a `RuntimeException` is passed to `onResult`.

Interestingly, `CompletionHandler.failed(Throwable exc, A attachment)` used in `AsynchronousSocketChannel.read` also accepts a `Throwable` instead of accepting an `Exception`, which I fail to find a justification for. Did this fact affected the signature of our callbacks?

Your question made me realize, that `result` may generally speaking also be an `Error`, and I do not handle this. Though if this is the case, then the code that completed the callback must have violated the recommendation from the `Error` class to not catch those. I see three ways of how I can proceed:

1. [handle `Throwable`s] Also handle situations when `result` is an `Error` (we must not wrap it into an `Exception` when handling). This approach works around (not a good thing in this case) a potential problem (code catching `Error`s and completing callbacks with them) which we may not have. But if we do have it, then this approach further hides the problem.
2. [handle only `Exception`s] Leave the idea mostly as is (handles `Exception`s but not `Error`s), do the following modifications
   2.1. wrap checked exceptions in `MongoException` instead of wrapping in `RuntimeException`;
   2.2. add `assertFalse(result instanceof Error)` to make the assumption about `Error`s explicit.
3. [handle only `RuntimeException`s] Add `assertTrue(result instanceof RuntimeException)` and then treat `result` as a `RuntimeException`, wrapping it in `MongoException` if needed.

I reject the option 1. I think that the option 2 is the least incorrect one, but I will not argue against the option 3. @jyemin Do you prefer 2 or 3, or maybe you see an even better approach that I missed?

Reply:
> The best answer to the question I can give is: result may not be an instance of RuntimeException in the scenarios in which not a RuntimeException is passed to onResult.

Yes, I see that now.  
> Did this fact affected the signature of our callbacks?

We were probably influenced more by Reactive Streams, which has `org.reactivestreams.Subscriber#onError(Throwable t)`.  Every async API that I've seen in Java does something like this (also see `Mono#onError` in Project Reactor).

In terms of what we should do, could we do the thing that doesn't change the current behavior?  I know you don't like it, but I'd rather not change error handling behavior in the scope of this ticket.

Reply:
Done.